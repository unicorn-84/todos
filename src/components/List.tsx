import React from 'react';
import clsx from 'clsx';
import { Item } from '.';
import type { ITodo, ITodoChangeMethods } from '../types';

interface IListProps
  extends Pick<ITodoChangeMethods, 'removeTodo' | 'toggleTodo'>,
  React.HTMLAttributes<HTMLUListElement> {
  todos: ITodo[]
}

const List: React.FC<IListProps> = ({
  todos,
  removeTodo,
  toggleTodo,
  className,
  ...props
}) => {
  return (
    <ul className={clsx('space-y-4 lg:space-y-6', className)} {...props}>
      {todos.map(({ todoId, task, complete }) => (
        <Item
          key={todoId}
          task={task}
          complete={complete}
          todoId={todoId}
          removeTodo={removeTodo}
          toggleTodo={toggleTodo}
        />
      ))}
    </ul>
  );
};

export default List;
