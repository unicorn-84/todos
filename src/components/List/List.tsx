import React from 'react';
import { Item } from '..';
import type { ITask } from '../../types';

interface IListProps extends React.HTMLAttributes<HTMLUListElement> {
  tasks: ITask[];
  toggleTask: (id: number) => void;
  removeTask: (id: number) => void;
}

const List: React.FC<IListProps> = ({
  tasks,
  removeTask,
  toggleTask,
  ...props
}) => {
  return (
    tasks.length > 0 && (
      <ul className="space-y-4 lg:space-y-6" {...props}>
        {tasks.map((task) => (
          <li key={task.id}>
            <Item task={task} removeTask={removeTask} toggleTask={toggleTask} />
          </li>
        ))}
      </ul>
    )
  );
};

export default List;
