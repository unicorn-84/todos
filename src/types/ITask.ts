interface ITask {
  taskId: number;
  taskTitle: string;
  taskComplete: boolean;
}

export default ITask;
