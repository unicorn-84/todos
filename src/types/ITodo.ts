interface ITodo {
  todoId: number;
  task: string;
  complete: boolean;
}

export default ITodo;
