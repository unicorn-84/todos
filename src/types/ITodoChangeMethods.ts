interface ITodoChangeMethods {
  addTodo: () => void;
  removeTodo: (id: number) => void;
  toggleTodo: (id: number) => void;
}

export default ITodoChangeMethods;
