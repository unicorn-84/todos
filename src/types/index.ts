export type { default as ITask } from './ITask';
