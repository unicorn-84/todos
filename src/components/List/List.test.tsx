import React from 'react';
import { render, screen } from '@testing-library/react';
import List from '.';
import { type ITask } from '../../types';

const tasks: ITask[] = [
  {
    id: 1,
    title: 'Task1',
    complete: true,
  },
  {
    id: 2,
    title: 'Task2',
    complete: false,
  },
];

describe('List', () => {
  test('should render correctly', () => {
    render(<List tasks={tasks} toggleTask={vi.fn()} removeTask={vi.fn()} />);

    expect(screen.getByRole('list')).toMatchSnapshot();
  });

  test('should render the list of tasks', () => {
    render(<List tasks={tasks} toggleTask={vi.fn()} removeTask={vi.fn()} />);

    expect(screen.getAllByRole('listitem')).toHaveLength(2);
  });

  test('should not render the empty list', () => {
    render(<List tasks={[]} toggleTask={vi.fn()} removeTask={vi.fn()} />);

    expect(screen.queryByRole('list')).not.toBeInTheDocument();
  });
});
