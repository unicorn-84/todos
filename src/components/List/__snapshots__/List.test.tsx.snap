// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List > should render correctly 1`] = `
<ul
  class="space-y-4 lg:space-y-6"
>
  <li>
    <div
      class="flex items-center justify-between space-x-3 lg:space-x-4 opacity-30"
    >
      <label
        class="label cursor-pointer w-full justify-start space-x-3 lg:space-x-4 p-0"
      >
        <input
          class="checkbox checkbox-primary checkbox-md lg:checkbox-lg"
          type="checkbox"
        />
        <span
          class="text-3xl lg:text-4xl break-words leading-none line-through"
        >
          Task1
        </span>
      </label>
      <button
        class="btn btn-warning btn-xs lg:btn-sm btn-square"
      >
        <svg
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 18L18 6M6 6l12 12"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </li>
  <li>
    <div
      class="flex items-center justify-between space-x-3 lg:space-x-4"
    >
      <label
        class="label cursor-pointer w-full justify-start space-x-3 lg:space-x-4 p-0"
      >
        <input
          class="checkbox checkbox-primary checkbox-md lg:checkbox-lg"
          type="checkbox"
        />
        <span
          class="text-3xl lg:text-4xl break-words leading-none"
        >
          Task2
        </span>
      </label>
      <button
        class="btn btn-warning btn-xs lg:btn-sm btn-square"
      >
        <svg
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 18L18 6M6 6l12 12"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </li>
</ul>
`;
